### PR TITLE
chore(wallet): Migrate Report Tx Sent P3A question into core

### PR DESCRIPTION
**Pull request by StephenHeaps**
Resolves https://github.com/brave/brave-browser/issues/36156

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

#### Desktop/Android:
0. Start browser with `--p3a-count-wallet-test-networks` flag. Ensure profile contains wallet with testnet tokens.
1. Ensure metric does not exist in local state.
2. Send transaction, ensure metric has value of 1.
3. Close browser, move system time ahead by seven days.
4. Open browser, ensure metric has value of 0.
5. Send another transaction, ensure metric has value of 1.

#### iOS:

1. Start browser with `Brave Wallet Count Test Networks` Brave Core Switch enabled. Ensure profile contains wallet with testnet tokens.
    - This can be enabled on iOS in main browser via ... -> Settings -> BraveCore Switches -> enable `Brave Wallet Count Test Networks`. You will have to quit the app and restart for changes to take effect.
2. Ensure metric does not exist in local state.
    - On iOS, local state can be viewed in `...` -> `Settings` -> `View Chromium Local State`
3. In Wallet, tap `...` and tap `Wallet Settings` -> `Networks` -> enable `Show Test Networks` switch.
4. Send transaction, ensure metric has value of 1.
5. Close browser, move system time ahead by seven days.
6. Open browser, ensure metric has value of 0.
7. Send another transaction, ensure metric has value of 1.

### CODE REVIEW

File: components/brave_wallet/browser/brave_wallet_p3a.cc [hunk 1]
```diff
@@ -22,6 +22,8 @@
 #include "components/prefs/pref_service.h"
 #include "components/prefs/scoped_user_pref_update.h"
 
+#include "base/command_line.h"
+
```
Comment (supermassive) [inline components/brave_wallet/browser/brave_wallet_p3a.cc:26]: nit: no empty line, so include gets sorted

File: components/brave_wallet/browser/brave_wallet_p3a.h [hunk 2]
```diff
@@ -119,12 +128,16 @@ class BraveWalletP3A : public KeyringServiceObserverBase,
...
   raw_ptr<BraveWalletService> wallet_service_;
   raw_ptr<KeyringService> keyring_service_;
+  raw_ptr<TxService> tx_service_;
   raw_ptr<PrefService> profile_prefs_;
   raw_ptr<PrefService> local_state_;
 
   mojo::Receiver<brave_wallet::mojom::KeyringServiceObserver>
       keyring_service_observer_receiver_{this};
 
+  mojo::Receiver<brave_wallet::mojom::TxServiceObserver>
```
Comment (supermassive) [inline components/brave_wallet/browser/brave_wallet_p3a.h:138]: Don't need `brave_wallet::` here and above

File: components/brave_wallet/browser/brave_wallet_p3a.cc [hunk 3]
```diff
@@ -406,4 +412,51 @@ void BraveWalletP3A::WalletCreated() {
...
+          chain_id == mojom::kFilecoinTestnet;
+      if (!count_test_networks && is_test_network) {
+        // Enable `--p3a-count-wallet-test-networks` flag for test network
+        // support
+        return;
+      }
+      auto tx_coin = is_eth_tx   ? mojom::CoinType::ETH
+                     : is_sol_tx ? mojom::CoinType::SOL
+                                 : mojom::CoinType::FIL;
+      ReportTransactionSent(tx_coin, true);
```
Comment (supermassive) [inline components/brave_wallet/browser/brave_wallet_p3a.cc:457]: Let's drop BraveWalletP3A.ReportTransactionSent form mojom

File: components/brave_wallet/browser/brave_wallet_p3a.cc [hunk 4]
```diff
@@ -406,4 +412,51 @@ void BraveWalletP3A::WalletCreated() {
...
+          chain_id == mojom::kFilecoinEthereumTestnetChainId ||
+          chain_id == mojom::kSolanaTestnet ||
+          chain_id == mojom::kSolanaDevnet ||
+          chain_id == mojom::kFilecoinTestnet;
+      if (!count_test_networks && is_test_network) {
+        // Enable `--p3a-count-wallet-test-networks` flag for test network
+        // support
+        return;
+      }
+      auto tx_coin = is_eth_tx   ? mojom::CoinType::ETH
```
Comment (supermassive) [inline components/brave_wallet/browser/brave_wallet_p3a.cc:454]: I'd suggest to rearrange the code a bit:
- early return with `if (tx_status != mojom::TransactionStatus::Approved) { return; }`
- extract `GetCoinTypeFromTxDataUnion` from `tx_service.cc` into `common_utils.cc` for example.
- get coin from `tx_data_union` using call above, early return for BTC and ZEC(I'll do them myself as they lack transition to Approved state now)
- depending coin check if `tx_type` is suitable for that coin. 
- also do testnet check per each coin separately. So, for example, `mojom::kFilecoinTestnet` is checked on for FIL
- the last line is `ReportTransactionSent(tx_coin, true);`

File: components/brave_wallet/browser/brave_wallet_p3a.cc [hunk 5]
```diff
@@ -406,4 +412,57 @@ void BraveWalletP3A::WalletCreated() {
...
 
+void BraveWalletP3A::OnTransactionStatusChanged(
+    mojom::TransactionInfoPtr tx_info) {
+  auto tx_status = tx_info->tx_status;
+  if (tx_status != mojom::TransactionStatus::Approved) {
+    return;
+  }
+  auto tx_coin = GetCoinTypeFromTxDataUnion(*tx_info->tx_data_union);
+  if (tx_coin == mojom::CoinType::BTC || tx_coin == mojom::CoinType::ZEC) {
+    // BTC, ZEV transactions are not tracked yet
```
Comment (supermassive) [inline components/brave_wallet/browser/brave_wallet_p3a.cc:423]: nit: ZEC

File: browser/brave_wallet/keyring_service_unittest.cc [hunk 6]
```diff
@@ -1967,7 +1977,8 @@ TEST_F(KeyringServiceUnitTest, RestoreLegacyBraveWallet) {
   KeyringService service(json_rpc_service(), GetPrefs(), GetLocalState());
   BraveWalletService brave_wallet_service(
       shared_url_loader_factory(), nullptr, &service, json_rpc_service(),
-      nullptr, nullptr, nullptr, GetPrefs(), GetLocalState(), false);
+      tx_service(&service).get(), nullptr, nullptr, GetPrefs(), GetLocalState(),
```
Comment (supermassive) [inline browser/brave_wallet/keyring_service_unittest.cc:1980]: This creates `unique_ptr<TxService>` and immediately destroys it leaving danglig pointer in `brave_wallet_service.tx_service_` -- most likely will crash in ASAN builds. It should be either TxService living on stack like `KeyringService service` above or allowing tx_service to be null for tests(see `CHECK_IS_TEST()`)

File: browser/brave_wallet/keyring_service_unittest.cc [hunk 7]
```diff
@@ -170,6 +171,15 @@ class KeyringServiceUnitTest : public testing::Test {
 
   JsonRpcService* json_rpc_service() { return json_rpc_service_; }
 
+  std::unique_ptr<TxService> tx_service(KeyringService* service) {
```
Comment (supermassive) [inline browser/brave_wallet/keyring_service_unittest.cc:174]: `tx_service()` naming style is for trivial getters only. For this case it should be like `MakeTxService(...)`

File: browser/brave_wallet/brave_wallet_p3a_unittest.cc [hunk 8]
```diff
@@ -38,18 +51,200 @@ class BraveWalletP3AUnitTest : public testing::Test {
...
   std::unique_ptr<base::HistogramTester> histogram_tester_;
   raw_ptr<KeyringService> keyring_service_;
+  raw_ptr<TxService> tx_service_;
   raw_ptr<BraveWalletService> wallet_service_;
   raw_ptr<BraveWalletP3A> wallet_p3a_;
+  raw_ptr<JsonRpcService> json_rpc_service_;
+  scoped_refptr<network::SharedURLLoaderFactory> shared_url_loader_factory_;
+  network::TestURLLoaderFactory url_loader_factory_;
+  data_decoder::test::InProcessDataDecoder in_process_data_decoder_;
+  std::string tx_hash1_;
```
Comment (supermassive) [inline browser/brave_wallet/brave_wallet_p3a_unittest.cc:247]: It seems we don't need `tx_hash1_` being a field

File: browser/brave_wallet/keyring_service_unittest.cc (hunk 6 above)
Comment (StephenHeaps) [reply browser/brave_wallet/keyring_service_unittest.cc]: TIL about `CHECK_IS_TEST` 👍 . I went with `MakeTxService` route given [check_is_test.h documentation](https://source.chromium.org/chromium/chromium/src/+/main:base/check_is_test.h;l=14-15?q=CHECK_IS_TEST).

File: browser/brave_wallet/brave_wallet_p3a_unittest.cc [hunk 9]
```diff
@@ -590,7 +589,7 @@ TEST_F(BraveWalletP3AUnitTest, SolTransactionSentObservation) {
       mojom::TxDataUnion::NewSolanaTxData(std::move(solana_tx_data)),
       mojom::kSolanaMainnet, sol_from(), &tx_meta_id));
 
-  tx_hash1_ =
+  std::string tx_hash1_ =
```
Comment (supermassive) [inline browser/brave_wallet/brave_wallet_p3a_unittest.cc:592]: Nit: local variables should not have trailing underscore

**Review by supermassive — APPROVED** [review]
Wallet core lgtm.
Please uplift to 1.64 and 1.65

**Review by DJAndries — APPROVED** [review]

**Review by yrliou — APPROVED** [review]
Just unblocking for wallet-core since @supermassive already approved.

**Review by josheleonard — APPROVED** [review]
Wallet Desktop Frontend ++

**Review by simoarpe — APPROVED** [review]
👍 Android code looks good! 
Thank you for the fix.

**Review by nuo-xu — APPROVED** [review]

**Comment by GeetaSarvadnya** [comment]
Verification **INPROGRESS** on

```
Brave | 1.65.73 Chromium: 123.0.6312.29 (Official Build) nightly (64-bit)
-- | --
Revision | 9cc34b8cfe9e8df8faede773545252783eefd953
OS | Windows 10 Version 22H2 (Build 19045.4046)
```
- Verified the test plan from https://github.com/brave/brave-core/pull/22360

**Pull request merged**